### PR TITLE
Feature/init

### LINKED ISSUE #1: Add a nice README

### LINKED ISSUE #2: Create a TYPO3 distribution for the trainings
https://github.com/typo3-trainer-network/typo3-distribution

### PR DESCRIPTION
**Pull request by DanielSiepmann**
Resolved: #1
Resolved: #2

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# For now, as extensions are in dev mode
+/composer.lock
+/var/*
+!/var/sqlite/
+/vendor
+/public/*
```
Comment (oliverklee) [inline .gitignore:1]: I propose sorting the `.gitignore` entries by name.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# TYPO3 Trainer Network Distribution
+
+This TYPO3 distribution contains an example TYPO3 installation.
+This installation includes all Extensions and examples used during official Trainings.
```
Comment (oliverklee) [inline README.md]: Please lowercase "extensions" and "trainings".

File: README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+## Installation
+
+1. Run `composer install`
+
+2. Run `TYPO3_CONTEXT=Development php -S 127.0.0.1:8080 -t public`
```
Comment (oliverklee) [inline README.md:9]: No blank line between list items, please.

Also, please always use `1.`. MarkDown will automatically increase the number. This reduces the work when lines are switched, added or removed.

File: TsLint.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+sniffs:
```
Comment (oliverklee) [inline TsLint.yml]: Is this the default name for the configuration file? If not; I'd prefer it to be placed in `config/`.

File: composer.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
+{
+    "name": "ttn/typo3-distribution",
+    "description": "Example distribution by the TYPO3 Trainer Network",
+    "type": "project",
+    "keywords": [
+        "typo3",
+        "example"
+    ],
+    "license": "GPL-2.0-or-later",
```
Comment (oliverklee) [inline composer.json:9]: Please also add a `LICENSE` file.

File: composer.json [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    "repositories": [
+        {
+            "type": "path",
+            "url": "localPackages/*"
+        }
+    ],
+    "require": {
+        "php": "~7.2 || ~7.3 || ~7.4",
+        "helmich/typo3-typoscript-lint": "^2.0",
+        "squizlabs/php_codesniffer": "^3.5",
```
Comment (oliverklee) [inline composer.json]: These two should be in require-dev.

Comment (oliverklee) [inline composer.json]: Also, for the dev tools, I recommend using the minor version so we don't get the bugs of older versions.

File: composer.json [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        "ttn/books": "^1.0",
+        "ttn/site-developer": "*",
+        "ttn/tea": "^2.0",
```
Comment (oliverklee) [inline composer.json]: For the example extensions, I'm for `@dev` so we don't need to release that often.

File: composer.json [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        "typo3/cms-backend": "*",
+        "typo3/cms-core": "^9.5",
+        "typo3/cms-frontend": "*",
+        "typo3/cms-install": "*",
+        "typo3/cms-lowlevel": "*"
```
Comment (oliverklee) [inline composer.json]: For the CMS-related packaged, please let's always use a real version like `^9.5`, not the death star.

File: composer.json [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    },
+    "require-dev": {
+        "helhum/typo3-console": "^5.7",
+        "typo3/cms-fluid-styled-content": "*",
+        "typo3/cms-tstemplate": "*"
+    },
+    "replace": {
+        "typo3-ter/typo3-distribution": "self.version"
```
Comment (oliverklee) [inline composer.json]: Please let's not do this for this package. If that extension existed, it should be helhum's dist, not our example dist.

File: composer.json [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    },
+    "autoload": {
+        "psr-4": {
+            "TTN\\Typo3Distribution\\": "Classes/"
+        }
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "TTN\\Typo3Distribution\\Tests\\": "Tests/"
```
Comment (oliverklee) [inline composer.json]: I don't think we have these two folders here AFAICS.

File: config/sites/developer/config.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    enabled: true
+    languageId: '0'
+    base: /
+    typo3Language: default
+    locale: en_US.UTF-8
+    iso-639-1: en
+    navigationTitle: English
+    hreflang: en-us
+    direction: ltr
+    flag: global
```
Comment (oliverklee) [inline config/sites/developer/config.yaml:16]: Please also let's have additional languages (either now or in a separate ticket).

File: localPackages/site_developer/Configuration/TCA/Overrides/sys_template.php [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+(function (string $extensionKey = 'site_developer') {
```
Comment (oliverklee) [inline localPackages/site_developer/Configuration/TCA/Overrides/sys_template.php]: Is there some kind of convention how to name the folder with the local packages? (Helmut uses `packages` IIRC.)

File: localPackages/site_developer/Configuration/TypoScript/Setup/Frontend/Page.typoscript [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    meta {
+        viewport = width=device-width, initial-scale=1
+    }
+
+    10 = FLUIDTEMPLATE
+    10 {
+        templateName = Default
+
+        layoutRootPaths {
+            10 = EXT:site_developer/Resources/Private/Layouts/Page/
```
Comment (oliverklee) [inline localPackages/site_developer/Configuration/TypoScript/Setup/Frontend/Page.typoscript]: I propose to name the site package `site` - that way we avoid the underscore, and the package name then also makes sense for TCCI and TCCE trainings.

File: localPackages/site_developer/Resources/Private/Layouts/Page/Default.html [hunk 14]
```diff
@@ -0,0 +1,27 @@
+<html xmlns:f="http://typo3.org/ns/TYPO3/CMS/Fluid/ViewHelpers"
+    data-namespace-typo3-fluid="true">
```
Comment (oliverklee) [inline localPackages/site_developer/Resources/Private/Layouts/Page/Default.html]: Would this also fit in one line? (Ditto in the other files with the same snippet.)

File: localPackages/site_developer/Resources/Private/Layouts/Page/Default.html [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+    <div class="header d-flex flex-column flex-md-row align-items-center p-3 px-md-4 mb-3 border-bottom shadow-sm">
+        <h5 class="my-0 mr-md-auto font-weight-normal"><a href="{menu.0.link}" class="text-dark">{menu.0.title}</a></h5>
```
Comment (oliverklee) [inline localPackages/site_developer/Resources/Private/Layouts/Page/Default.html]: Please also let's use indentation here, i.e., the A tag should go on a separate line, and its contents on another line. Ditto for the H1, P etc. below.

File: localPackages/site_developer/Resources/Private/Templates/Page/Default.html [hunk 16]
```diff
@@ -0,0 +1,9 @@
+<html xmlns:f="http://typo3.org/ns/TYPO3/CMS/Fluid/ViewHelpers"
+    data-namespace-typo3-fluid="true">
+
+    {f:layout(name: 'Default')}
```
Comment (oliverklee) [inline localPackages/site_developer/Resources/Private/Templates/Page/Default.html]: Why the inline notation here?

File: README.md [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+## Credentials
+
+Administrator
+    username: `admin`
+    password: `admin`
+
+Install Tool Password
+    `itp`
```
Comment (oliverklee) [inline README.md:20]: We also need to mention the license of the Bootstrap code.

File: localPackages/site_developer/composer.json [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+            "email": "trainer-network@typo3.org"
+        }
+    ],
+    "support": {
+        "issues": "https://github.com/typo3-trainer-network/typo3-distribution/issues",
+        "source": "https://github.com/typo3-trainer-network/typo3-distribution"
+    },
+    "require": {
+        "php": "~7.2 || ~7.3 || ~7.4",
+        "typo3/cms-core": "^9.5"
```
Comment (oliverklee) [inline localPackages/site_developer/composer.json]: Also Fluid, frontend and FSC AFAICS.

File: localPackages/site_developer/composer.json [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    "autoload": {
+        "psr-4": {
+            "TTN\\SiteDeveloper\\": "Classes/"
+        }
+    },
+    "autoload-dev": {
+        "psr-4": {
+            "TTN\\SiteDeveloper\\Tests\\": "Tests/"
+        }
+    },
```
Comment (oliverklee) [inline localPackages/site_developer/composer.json]: There is no code here, hence there should be no autoloading.

File: localPackages/site_developer/composer.json [hunk 20]
```diff
@@ -0,0 +1,40 @@
+{
+    "name": "ttn/site-developer",
+    "description": "Sitepackage for developer training",
```
Comment (oliverklee) [inline localPackages/site_developer/composer.json]: Sitepackage -> Site package

developer training -> TYPO3 Trainer Network trainings

File: localPackages/site_developer/ext_emconf.php [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    'author_company' => 'TYPO3 Trainer Network',
+    'autoload' => [
+        'psr-4' => [
+            'TTN\\SiteDeveloper\\' => 'Classes/',
+        ],
+    ],
+    'autoload-dev' => [
+        'psr-4' => [
+            'TTN\\SiteDeveloper\\Tests\\' => 'Tests/',
+        ],
```
Comment (oliverklee) [inline localPackages/site_developer/ext_emconf.php]: Ditto.

File: localPackages/site_developer/ext_emconf.php [hunk 22]
```diff
@@ -0,0 +1,27 @@
+<?php
+$EM_CONF[$_EXTKEY] = [
+    'title' => 'Developer Site',
+    'description' => 'Sitepackage for developer training',
```
Comment (oliverklee) [inline localPackages/site_developer/ext_emconf.php]: Ditto.

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: composer.json (hunk 5 above)
Comment (DanielSiepmann) [reply composer.json:9]: This already exists since your first initial commit.

File: localPackages/site_developer/composer.json (hunk 20 above)
Comment (DanielSiepmann) [reply localPackages/site_developer/composer.json]: I followed https://docs.typo3.org/m/typo3/docs-how-to-document/master/en-us/GeneralConventions/ContentStyleGuide.html#spelling-preferred-terms-reference
where we defined "sitepackage" as a special name within TYPO3:

> sitepackage
    In the context of a TYPO3 sitepackage, a TYPO3 sitepackage extension or the sitepackage builder, we use the spelling sitepackage and not site package.

Should I change that?

File: TsLint.yml (hunk 4 above)
Comment (DanielSiepmann) [reply TsLint.yml]: Could do that. `config` is a special folder setup by TYPO3. Right now there are site configurations.

File: .gitignore (hunk 1 above)
Comment (DanielSiepmann) [reply .gitignore:1]: I've group sorted. Some parts are excluded and would be sorted separate as they are prefixed with `!`.

File: localPackages/site_developer/composer.json (hunk 18 above)
Comment (DanielSiepmann) [reply localPackages/site_developer/composer.json]: Until now I've not added those dependencies to the sitepackages again, as they are already part of the main package, and sitepackaes only live in this context.

Should I still add them, or even remove `typo3/cms-core`?

File: config/sites/developer/config.yaml (hunk 11 above)
Comment (DanielSiepmann) [reply config/sites/developer/config.yaml:16]: Created issue #4

File: localPackages/site_developer/Configuration/TypoScript/Setup/Frontend/Page.typoscript (hunk 13 above)
Comment (DanielSiepmann) [reply localPackages/site_developer/Configuration/TypoScript/Setup/Frontend/Page.typoscript]: Thought to use separate sites and site packages within the same installation.
This way we would prevent one group from changing stuff in other areas.

Should I adjust it to `site`, and should be keep some separation? Right now the pages in the database also are intended for developer training only, with a dedicated page root.

File: localPackages/site_developer/Resources/Private/Templates/Page/Default.html (hunk 16 above)
Comment (DanielSiepmann) [reply localPackages/site_developer/Resources/Private/Templates/Page/Default.html]: I always use inline when the result is not a tag, and no area is necessary, which would be necessary when providing a fallback for e.g. an `f:section`, but of course I can change that.

File: composer.json (hunk 8 above)
Comment (DanielSiepmann) [reply composer.json]: We always use the version for `typo3/cms-core` and `*` for everything else. As they are only compatible with the exact same version, we only need to change one thing.

But I've adjusted it here.

File: localPackages/site_developer/Configuration/TCA/Overrides/sys_template.php (hunk 12 above)
Comment (DanielSiepmann) [reply localPackages/site_developer/Configuration/TCA/Overrides/sys_template.php]: I'm not aware of any. That's just our convention.

**Comment by DanielSiepmann** [comment]
@oliverklee please can you also check whether everything works locally as expected?

File: localPackages/site_developer/composer.json (hunk 18 above)
Comment (oliverklee) [reply localPackages/site_developer/composer.json]: The packages that are for the BE and the installation itself (core etc.) should be required in the main `composer.json`, and the packages required for the FE rendering should be required in the site package's `composer.json` (even if there are duplicate entries).

File: localPackages/site_developer/composer.json (hunk 20 above)
Comment (oliverklee) [reply localPackages/site_developer/composer.json]: Okay, then let's keep it that way.

File: README.md [hunk 23]
```diff
@@ -0,0 +1,24 @@
+# TYPO3 Trainer Network Distribution
+
+This TYPO3 distribution contains an example TYPO3 installation.
+This installation includes all extensions and examples used during official trainings.
+
+## Installation
+
+1. Run `composer install`
+1. Run `TYPO3_CONTEXT=Development php -S 127.0.0.1:8080 -t public`
```
Comment (oliverklee) [inline README.md]: Please also add a next step with links to the BE and FE.

**Review by oliverklee — CHANGES_REQUESTED** [review]
I've tested this. The installation works perfectly! Thanks a lot!

Could you also please add the `composer.lock`? Thanks!

**Review by oliverklee — CHANGES_REQUESTED** [review]
And there also are PHPCodeSniffer warnings that need fixing.

**Review by oliverklee — APPROVED** [review]

**Pull request merged**